pkg/gatewayserver/io/basicstationlns/messages/upstream.go

```diff
@@ -0,0 +1,294 @@
+	up.ReceivedAt = time.Now()
+
+	parsedMHDR := ttnpb.MHDR{}
+	err := lorawan.UnmarshalMHDR([]byte{byte(req.MHdr)}, &parsedMHDR)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	micBytes, err := getInt32AsByteSlice(req.MIC)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+	up.Payload = &ttnpb.Message{
+		MHDR: parsedMHDR,
+		MIC:  micBytes,
+		Payload: &ttnpb.Message_JoinRequestPayload{JoinRequestPayload: &ttnpb.JoinRequestPayload{
+			JoinEUI:  req.JoinEUI.EUI64,
+			DevEUI:   req.DevEUI.EUI64,
+			DevNonce: [2]byte{byte(req.DevNonce), byte(req.DevNonce >> 8)},
+		}},
+	}
+
+	up.RawPayload, err = lorawan.MarshalMessage(*up.Payload)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	timestamp := uint32(req.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
+
+	ulToken := ttnpb.UplinkToken{
+		GatewayAntennaIdentifiers: ttnpb.GatewayAntennaIdentifiers{
+			GatewayIdentifiers: ids,
+			AntennaIndex:       uint32(req.RadioMetaData.UpInfo.RCtx),
+		},
+		Timestamp: timestamp,
+	}
+	ulTokenBytes, err := ulToken.Marshal()
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	rxTime := time.Unix(req.RadioMetaData.UpInfo.RxTime, 0)
+	rxMetadata := &ttnpb.RxMetadata{
+		GatewayIdentifiers: ids,
+		Time:               &rxTime,
+		Timestamp:          timestamp,
+		RSSI:               req.RadioMetaData.UpInfo.RSSI,
+		SNR:                req.RadioMetaData.UpInfo.SNR,
+		UplinkToken:        ulTokenBytes,
+	}
+	up.RxMetadata = append(up.RxMetadata, rxMetadata)
+
+	loraDR, err := getDataRateFromIndex(bandID, req.RadioMetaData.DataRate)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+	up.Settings = ttnpb.TxSettings{
+		Frequency:     req.RadioMetaData.Frequency,
+		DataRateIndex: ttnpb.DataRateIndex(req.RadioMetaData.DataRate),
+		DataRate:      loraDR,
+	}
+
+	return up, nil
+}
+
+// ToUplinkMessage extracts fields from the basic station Join Request "jreq" message and converts them into an UplinkMessage for the network server.
+func (updf *UplinkDataFrame) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (ttnpb.UplinkMessage, error) {
+	up := ttnpb.UplinkMessage{}
+	up.ReceivedAt = time.Now()
+
+	parsedMHDR := ttnpb.MHDR{}
+	err := lorawan.UnmarshalMHDR([]byte{byte(updf.MHdr)}, &parsedMHDR)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+	if (parsedMHDR.MType != ttnpb.MType_UNCONFIRMED_UP) && (parsedMHDR.MType != ttnpb.MType_CONFIRMED_UP) {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame
+	}
+
+	micBytes, err := getInt32AsByteSlice(updf.MIC)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errUplinkDataFrame.WithCause(err)
+	}
+
+	var fport uint32
+	if updf.FPort == -1 {
+		fport = 0
+	} else {
+		fport = uint32(updf.FPort)
+	}
+
+	devAddrBytes, err := getInt32AsByteSlice(updf.DevAddr)
```

Comment:
Is this really shorter/more concise than inlining this func?

Reply:
Short functions are inlined by the compiler during code generation but at the source level I find this a bit more readable. I'm fine with changing it if needed.